include/vata2/simutil/binary_relation.hh

```diff
@@ -0,0 +1,935 @@
+		{
+			size_t j = 0;
+			while ((j < head.size()) && !this->sym(i, head[j]))
+			{
+				++j;
+			}
+
+			if (j < head.size())
+			{
+				index[i] = j;
+			}
+			else
+			{
+				index[i] = head.size();
+				head.push_back(i);
+			}
+		}
+	}
+
+	// and composition
+	BinaryRelation& operator&=(const BinaryRelation& rhs)
+	{
+		assert(size_ == rhs.size_);
+
+		for (size_t i = 0; i < size_; ++i)
+		{
+			for (size_t j = 0; j < size_; ++j)
+			{
+				this->set(i, j, this->get(i, j) && rhs.get(i,j));
+			}
+		}
+
+		return *this;
+	}
+
+	// transposition
+	BinaryRelation& transposed(BinaryRelation& dst) const
+	{
+		dst.resize(size_);
+		for (size_t i = 0; i < size_; ++i)
+		{
+			for (size_t j = 0; j < size_; ++j)
+			{
+				dst.set(j, i, this->get(i, j));
+			}
+		}
+
+		return dst;
+	}
+
+
+	/**
+	 * @brief  Creates a mapping from elements to their images
+	 *
+	 * This method maps elements to their images. Formally, it creates for each
+	 * element 'x' the mapping 'x -> Y' where 'Y = {y | xRy}'
+	 *
+	 * @param[out]  dst  The result mapping every element to its images
+	 *
+	 * @see  build_inv_index
+	 */
+	void build_index(IndexType& dst) const
+	{
+		dst.resize(size_);
+
+		auto rowStart = data_.begin();
+		auto src = rowStart;
+		auto end = src + size_*rowSize_;
+		auto dstIter = dst.begin();
+
+		for (; src != end; src = rowStart, ++dstIter)
+		{
+			auto rowEnd = src + size_;
+			rowStart = src + rowSize_;
+
+			for (size_t i = 0; src != rowEnd; ++src, ++i)
+			{
+				if (*src)
+				{
+					dstIter->push_back(i);
+				}
+			}
+		}
+	}
+
+	/**
+	 * @brief  Creates a mapping from elements to their co-images
+	 *
+	 * This method maps elements to their co-images. Formally, it creates for each
+	 * element 'x' the mapping 'x -> Y' where 'Y = {y | yRx}'
+	 *
+	 * @param[out]  dst  The result mapping every element to its co-images
+	 *
+	 * @see build_index
+	 */
+	void build_inv_index(IndexType& dst) const
+	{
+		dst.resize(size_);
+
+		for (size_t i = 0; i < size_; ++i)
+		{
+			for (size_t j = 0; j < size_; ++j)
+			{
+				if (this->get(i, j))
+				{
+					dst[j].push_back(i);
+				}
+			}
+		}
+	}
+
+	/**
+	 * @brief  Creates a mapping from elements to their images and co-images
+	 *
+	 * This method the call to build_index and build_inv_index, i.e. it maps
+	 * elements to their images (in @p ind) and their co-images (in @p inv).
+	 *
+	 * @param[out]  ind  The result mapping every element to its images
+	 * @param[out]  inv  The result mapping every element to its co-images
+	 *
+	 * @see build_index build_inv_index
+	 */
+	void build_index(IndexType& ind, IndexType& inv) const
+	{
+		ind.resize(size_);
+		inv.resize(size_);
+
+		for (size_t i = 0; i < size_; ++i)
+		{
+			for (size_t j = 0; j < size_; ++j)
+			{
+				if (this->get(i, j))
+				{
+					ind[i].push_back(j);
+					inv[j].push_back(i);
+				}
+			}
+		}
+	}
+
+	/**
+	 * @brief  Restricts the relation to its symmetric fragment
+	 */
+	void RestrictToSymmetric()
```

Comment:
Can't say I'm advocate for snake case of other case, but I would probably pick one and be consisent through the project.

Reply:
solved